src/toxcore/dht_new/client.rs

```diff
@@ -0,0 +1,125 @@
+
+/// peer info.
+#[derive(Clone, Debug)]
+pub struct Client {
+    /// Public key of peer
+    pub pk: PublicKey,
+    /// socket address of peer
+    pub addr: SocketAddr,
+    /// tx split of channel to send packet to this peer via udp socket
+    pub tx: Tx,
+    /// last sent ping_id to check PingResponse is correct
+    pub ping_id: u64,
+    // /// last sent NodesRequest id to check NodesResponse is correct
+    // pub nodes_id: u64,
+    // /// last sent NatPingRequest id to check NatPingResponse is correct
+    // pub nat_ping_id: u64,
+    /// precomputed key for this peer
+    pub precomputed_key: PrecomputedKey,
+}
+
+impl Client {
+    /// create Client object
+    pub fn new(precomputed_key: PrecomputedKey, pk: PublicKey, addr: SocketAddr, tx: Tx) -> Client {
+        Client {
+            pk: pk,
+            addr: addr,
+            tx: tx,
+            precomputed_key: precomputed_key,
+            ping_id: 0,
+            // nodes_id: 0,
+            // nat_ping_id: 0,
+        }
+    }
+    /// actual send method
+    fn send(&self, packet: DhtPacket) -> IoFuture<()> {
+        Box::new(self.tx.clone() // clone tx sender for 1 send only
+            .send((self.addr.clone(), packet))
+            .map(|_tx| ()) // ignore tx because it was cloned
+            .map_err(|e| {
+                // This may only happen if rx is gone
+                // So cast SendError<T> to a corresponding std::io::Error
+                error!("send to peer error {:?}", e);
+                Error::from(ErrorKind::UnexpectedEof)
+            })
+        )
+    }
+    /// respond with PingResponse to peer
+    pub fn send_ping_response(&self, resp_payload: PingResponsePayload) -> IoFuture<()> {
+        let ping_resp = DhtPacket::PingResponse(PingResponse::new(&self.precomputed_key.clone(), &self.pk, resp_payload));
+        debug!("PingResp made {:?}", ping_resp);
+        self.send(ping_resp)
+    }
+    /// respond with NodesResponse to peer
+    pub fn send_nodes_response(&self, resp_payload: NodesResponsePayload) -> IoFuture<()> {
+        let nodes_resp = DhtPacket::NodesResponse(NodesResponse::new(&self.precomputed_key.clone(), &self.pk, resp_payload));
+        debug!("NodesResp made {:?}", nodes_resp);
+        self.send(nodes_resp)
+    }
+    /// respond with NatPingResponse to peer
+    pub fn send_nat_ping_response(&self, rpk: &PublicKey, resp_payload: NatPingResponse) -> IoFuture<()> {
+        let payload = DhtRequestPayload::NatPingResponse(resp_payload);
+        let nat_ping_resp = DhtPacket::DhtRequest(DhtRequest::new(&self.precomputed_key.clone(), rpk, &self.pk, payload));
+        self.send(nat_ping_resp)
+    }
+    /// relay NatPingRequest/NatPingResponse to target peer
+    pub fn relay_nat_ping_packet(&self, addr: &SocketAddr, request: DhtRequest) -> IoFuture<()> {
+        let packet = DhtPacket::DhtRequest(request);
+        Box::new(self.tx.clone() // clone tx sender for 1 send only
```

Comment:
Create method `send_to` as method `send` but with explicit addr-argument. Method `send` should use `send_to` inside it passing `self.addr` as addr.